rfc/0084-app-service-level-telemetry.md

```diff
@@ -0,0 +1,517 @@
+latencies that when not serviced, lead to other processing.  This is worth
+mentioning because it will be important to compare a deployment to itself, and
+there may be limited value in comparing across deployments.
+
+## Capella Service Level Metrics
+
+The counters described above give a good picture of actual impact on a deployed
+app, but they do not give a sense of whether or not a given deployment is
+meeting an expected service level.  Said another way, a Couchbase Capella
+Cluster deployment could be operational, but very slow during a rebalance
+leading to user discontent, but never turn into an error.
+
+Service levels are specific to the user's targets.  While it's impossible to
+properly measure whether or not an end-user's app is meeting service level
+targets, we can establish a predefined histogram that would allow Couchbase to
+simply and inexpensively understand if the service level has likely changed.
+
+It's proposed that we establish a set of standardized histograms, specific to
+services and to the service level expected with the service using those
+parameters.  Counters for these histograms can be sent to Couchbase and both
+stored individually and aggregated.
+
+As an aggregated set of metrics, it would give Couchbase a rough view over the
+service level the user is experiencing.  Deviations in service level can be dug
+into and, in particular, we can check for correlation to Couchbase Capella
+management activities like upgrades.
+
+This list represents the services, buckets.
+
+* Data Service, Non-Durable
+  * Retrieval
+    * <1ms
+    * 1-10ms
+    * 10-100ms
+    * 100-500ms
+    * 500-1000ms
+    * 1s-2.5s
+    * \>2.5s
+  * Mutation
+    * <1ms
+    * 1-10ms
+    * 10-100ms
+    * 100-500ms
+    * 500-1000ms
+    * 1s-2.5s
+    * \>2.5s
+* Data Service, Durable
+  * Mutation
+    * 1-10ms
+    * 10-100ms
+    * 100-1000ms
+    * 1-2s
+    * 2s-5s
+    * 5s-10s
+    * \>10s
+* Query Service
+  * Query (all)
+    * 0-100ms
+    * 100-1000ms
+    * 1s-10s
+    * 10s-30s
+    * 30-75s
+    * \>75s
+* FTS Service
+  * Search (all)
+    * 0-100ms
+    * 100-1000ms
+    * 1s-10s
+    * 10s-30s
+    * 30-75s
+    * \>75s
+* Analytics Service
+  * Query (all)
+    * 0-100ms
+    * 100-1000ms
+    * 1s-10s
+    * 10s-30s
+    * 30-75s
+    * \>75s
+
+The Indexing Service and Eventing Services are not included as they are not
+network-available services for applications to make requests from.
+
+As with the Service Experience Counters above, it may be incorrect to try to
+compare between deployments.  For example, the query complexity may differ
+significantly or the data size and use of features like compression may vary
+significantly between two applications.
+
+# API and Implementation Details
+
+It should be possible to reuse existing Meter implementation, that according to
+RFC-67 (Extended SDK Observability) already builds histograms for request
+latencies.
```

Comment:
The meter infrastructure we have in the SDKs doesn't keep track of histograms in a way that we could reuse easily. Everything is contained inside the implementation of the meter interface, so the SDK is agnostic to what the meter does with what's passed on to it. Some implementations of the meter we provide, like the `LoggingMeter` builds histograms but it's all internal to it.

Reply:
My understanding is that the intention is to extend the existing default meter – i.e. the LoggingMeter (which wasn't what I had in mind in my previous comment above).

However, the metrics we currently record for the 'Operations Value Recorder' in RFC-78 are higher level than the ones we need for app telemetry:
> Instrumented at: when the response has been decoded, before it is marked as completed towards the user

Whereas in this RFC we have:
>  Each metric value increment represents a single network operation (e.g. GET_ALL_REPLICAS should be counted as 1 + number_of_replicas).

In the RFC-78 meter we would record a single metric value for a `get_all_replicas` operation, not `1 + number_of_replicas` that is required for app telemetry.

Reply:
Ah I see I've commented basically the same thing elsewhere.  Yes - it would be a lot easier if we regarded operations in the same way that RFC-78 does, so we can leverage the same existing instrumentation points.

Reply:
> In the RFC-78 meter we would record a single metric value for a get_all_replicas operation, not 1 + number_of_replicas that is required for app telemetry.
Here we collect metrics for the server, and from the server perspective there is no `get_all_replicas` operation, and we cannot assign any hostname to this operation. So no, we cannot account "virtual" operations, only physical commands, that are visible on the network and for the server.

Reply:
Updated in the RFC